decoder/vaapidecoder_mpeg2.cpp

```diff
@@ -400,36 +438,73 @@ YamiStatus VaapiDecoderMPEG2::processDecodeBuffer()
         break;
     case YamiParser::MPEG2::MPEG2_EXTENSION_START_CODE:
-        DEBUG("picture coding extension start code");
-        if (m_nextStartCode == next_code) {
-            if (!m_parser->parsePictureCodingExtension(m_stream.get())) {
+        extID = static_cast<YamiParser::MPEG2::ExtensionIdentifierType>(
+            m_stream->nalData[1] >> 4);
+        if (extID == YamiParser::MPEG2::kPictureCoding) {
+            if (m_nextStartCode == next_code) {
+                if (!m_parser->parsePictureCodingExtension(m_stream.get())) {
+                    return YAMI_DECODE_PARSER_FAIL;
+                }
+                m_pictureCodingExtension
+                    = m_parser->getPictureCodingExtension();
+
+                // picture can be created as soon as the first slice is
+                // parsed, but from here to there Extension Start Code can be
+                // parsed to update information like the Quant Matrix
+
+                m_canCreatePicture = true;
+                m_previousStartCode = m_nextStartCode;
+                m_nextStartCode = YamiParser::MPEG2::MPEG2_SLICE_START_CODE_MIN;
+                m_isParsingSlices = true;
+            }
+        }
+        else if (extID == YamiParser::MPEG2::kQuantizationMatrix) {
+            INFO("this is a new quant matrix");
+            if (!m_parser->parseQuantMatrixExtension(m_stream.get())) {
                 return YAMI_DECODE_PARSER_FAIL;
             }
-            m_pictureCodingExtension = m_parser->getPictureCodingExtension();
-
-            status = createPicture();
-            if (status != YAMI_SUCCESS) {
-                return status;
+            m_quantMatrixExtension = m_parser->getQuantMatrixExtension();
+            updateIQMatrix(&m_quantMatrixExtension->quantizationMatrices);
+            m_loadNewIQMatrix = true;
+        }
+        else if (extID == YamiParser::MPEG2::kSequence) {
+            if (!m_parser->parseSequenceExtension(m_stream.get())) {
+                return YAMI_DECODE_PARSER_FAIL;
             }
+            m_sequenceExtension = m_parser->getSequenceExtension();
             m_previousStartCode = m_nextStartCode;
-            m_nextStartCode = YamiParser::MPEG2::MPEG2_SLICE_START_CODE_MIN;
-            m_isParsingSlices = true;
+            m_nextStartCode = YamiParser::MPEG2::MPEG2_GROUP_START_CODE;
         }
         break;
     case YamiParser::MPEG2::MPEG2_SEQUENCE_END_CODE:
         INFO("End of sequence received");
-        status = decodePicture();
-        if (status != YAMI_SUCCESS) {
-            return status;
-        }
-        status = m_DPB.outputPreviousPictures(m_currentPicture, true);
-        if (status != YAMI_SUCCESS) {
-            return status;
+        if (m_isParsingSlices) {
+            status = decodePicture();
+            if (status != YAMI_SUCCESS) {
+                return status;
+            }
+            status = m_DPB.outputPreviousPictures(m_currentPicture, true);
```

Comment:
Ok, but I still have other related question. 1, 2(with close gop) and 3  should flush the reference frame in dpb, right? It will avoid frames in next gop reference previous gop. But I did not see reference frame clean logical in the code.